core/matrix/batch_diagonal_kernels.hpp

```diff
@@ -0,0 +1,128 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2022, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_MATRIX_BATCH_DIAGONAL_KERNELS_HPP_
+#define GKO_CORE_MATRIX_BATCH_DIAGONAL_KERNELS_HPP_
+
+
+#include <ginkgo/core/matrix/batch_diagonal.hpp>
+
+
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/batch_dense.hpp>
+#include <ginkgo/core/matrix/diagonal.hpp>
+
+
+namespace gko {
+namespace kernels {
+
+
+#define GKO_DECLARE_BATCH_DIAGONAL_APPLY_KERNEL(_type)      \
+    void apply(std::shared_ptr<const DefaultExecutor> exec, \
+               const matrix::BatchDiagonal<_type>* a,       \
+               const matrix::BatchDense<_type>* b,          \
+               matrix::BatchDense<_type>* x)
+
+#define GKO_DECLARE_BATCH_DIAGONAL_SIMPLE_APPLY_KERNEL(_type)      \
+    void simple_apply(std::shared_ptr<const DefaultExecutor> exec, \
+                      const matrix::BatchDiagonal<_type>* a,       \
+                      matrix::BatchDense<_type>* b)
```

Comment:
Is this being used anywhere ? I think this is an in-place apply ? I think you should name it that way.

Reply:
No, for now it's not being used. Would you say we should remove it? I kept it because at some point we may need to offer an interface to this too, something like `apply_in_place`.